### PR TITLE
fix(issue): [Bug]: gsd-auto-orchestrator terminal/pause event leaves sessions table status='running'; /gsd auto deadlocks indefinitely

### LINKED ISSUE #1490: [Bug]: gsd-auto-orchestrator terminal/pause event leaves sessions table status='running'; /gsd auto deadlocks indefinitely
## Summary

When `gsd-auto-orchestrator` reaches a `terminal / reason: pause` state mid-slice, the SQLite `sessions` row in `.gsd/gsd.db` is **not** demoted from `status='running'` to a terminal status. The journal (`journal/<date>.jsonl`) records the orchestrator stop correctly, the project snapshot reports `Blockers: None` and `Next: Execute T0X`, and the daemon process is healthy — but `/gsd auto` is rejected indefinitely with `Could not start GSD auto mode: Session already active for <project> (sessionId: 019f74c9-…, status: running)`. There is no user-visible way to recover short of restarting the daemon or hand-editing the SQLite row. The bug surfaced in M032/S02 immediately after `execute-task T01` finalized.

### Reproduction steps

1. Start auto-mode on a project whose `.gsd/` has at least one slice with multiple tasks.
   ```
   gsd auto --project <path>
   ```
2. Let auto-mode run through a complete unit (e.g. `execute-task S02/T01`) so the orchestrator finalizes it normally.
3. Cause the orchestrator to enter `terminal / reason: pause` after that unit finalizes (in our case, S02-T01 finalized and the orchestrator paused between T01 and T02 dispatch; the trigger condition in `gsd-auto-orchestrator` is opaque from the user side, but `journal/2026-07-18.jsonl` line seq 29 is the smoking gun).
4. Wait or attempt to resume.
5. Try to resume work on the next task:
   ```
   gsd auto --project <path>
   ```
6. **Expected:** Either auto-mode continues from `Next: Execute T02`, or `gsd auto` is rejected with a clear "paused" error and a recovery command.
7. **Actual:** `Could not start GSD auto mode: Session already active for /home/tumengchao/java/src/zerone-all (sessionId: 019f74c9-0a95-7a9a-b1a2-0d3d2be04398, status: running)` — the daemon is alive, the snapshot is healthy, but the next dispatch never happens. The `sessions` row's `status` column is stuck on `running` even though the journal shows the orchestrator is `terminal / reason: pause`.

### Environment

- `open-gsd/gsd-pi`: latest main (pulled 2026-07-18, version matched plugin's >=2.53,<3 contract)
- `open-gsd-hermes` plugin: v1.2.0
- `gsd-mcp-server` v1.5.0
- Hermes Agent gateway: 0.18.2 (linux-x64)
- OS: Linux 6.8.0-134-generic
- Project: `~/java/src/zerone-all`, .gsd/ on local disk

### Expected vs actual behavior (verbatim)

- `journal/.jsonl` seq 28 → seq 29:
  ```
  seq 28  "eventType":"orchestrator-dispatch-match","data":{"name":"advance","reason":"executing → execute-task","unitType":"execute-task","unitId":"M032/S02/T01"}
  seq 29  "eventType":"orchestrator-terminal","data":{"source":"auto-orchestrator","name":"stop","reason":"pause"}
  ```
- Subsequent `iteration-end` for T01 records `action: "next"`, but no new `iteration-start` ever fires for T02.
- `.gsd/runtime/units/execute-task-M032-S02-T01.json` reads:
  ```json
  {"phase":"finalized","lastProgressKind":"finalize-success","recoveryAttempts":0,"continueHereFired":false}
  ```
- `gsd.db` `sessions` row for `019f74c9-0a95-7a9a-b1a2-0d3d2be04398` still reads `status='running'`, `phase='execute'`, `updated_at=2026-07-18T11:31:27Z`.
- `gsd auto` rejects with the message quoted above; the daemon (PID confirmed alive, CPU ~0.6%, 0 child processes) silently idles.

## Proposed solution

The two status-tracking surfaces (append-only `journal/<date>.jsonl` and the SQLite `sessions` table) need a single source of truth for "is this session still dispatchable?" Right now, the orchestrator's `terminal` event only updates the journal; the SQLite row is left stale, and `gsd auto`'s `startSession` precheck is reading that stale row.

Concrete fixes (in order of surgical impact):

1. **Demote on terminal in the same transaction.** In `gsd-auto-orchestrator` (the orchestrator that emits the `terminal / reason: pause` event at `seq 29`), wrap the journal append in a transaction that also `UPDATE sessions SET status='paused' (or 'closed' depending on reason), updated_at=now WHERE id=?`. The MCP server's `startSession` guard will then either reject with a recoverable "paused" state or the next resume path will pick it up.

2. **Reconcile on `gsd auto` start.** Before the "session already active" guard fires, replay `journal/<date>.jsonl` for that sessionId: if the latest `orchestrator-terminal` event for the session is newer than the row's `updated_at`, re-derive the row's status from that event (`terminal/pause` → `paused`, `terminal/complete` → `closed`, `terminal/cancelled` → `cancelled`, etc.). This keeps the bug from persisting even when the orchestrator process is killed before step 1 can flush.

3. **Add a recovery tool.** A new `gsd session reconcile --session-id <id>` (or extend `gsd doctor --fix`) that runs the replay once and re-emits the journal→DB projection. This gives users a non-destructive escape hatch without requiring a daemon restart.

4. **Make the error actionable.** When `startSession` rejects with "Session already active ... status: running", include the project snapshot's `Blockers` and `Next` lines in the error message so the user can tell at a glance whether this is a stuck-row bug (the bug at hand) versus a genuinely active session.

All four changes are backward compatible. (3) is purely additive; (1) and (2) only change the `status` field write path, which is already inconsistent across versions.

## Alternatives considered

- **Restart the daemon (`kill -9 <gsd-pid>`).** This unblocks `/gsd auto` but loses any in-memory progress state and forces MCP server reconnection. Works as a one-off escape hatch, not a fix.
- **Hand-edit `gsd.db` `UPDATE sessions SET status='paused'`.** Works, but it's a privilege escalation for the user (manual SQL against a file the daemon holds). Should not be the recommended path.
- **Wait for session TTL / cleanup loop.** There is none in the observed behavior. The daemon never reaps the row.
- **Use `gsd recover --confirm`.** Triggers the known SHA-mismatch cascade (#1334 et al.) and re-plans 31 milestones. Categorically wrong tool for this case.
- **Close the session with `gsd session close`.** Not exposed to users — and even if it were, the orchestrator is in a paused state, not a closed one; closing would lose the resume path entirely.

## Use cases

- Long-running auto-mode over many tasks (e.g. M032 S02 has 4 tasks; M031 had 5; M029 had 7). Any orchestrator-pause between tasks now deadlocks the entire milestone until manual intervention.
- Unattended / cron-driven `gsd auto` runs: the user discovers the deadlock hours later via "still no progress" and has no clean way to recover without restarting infra.
- CI pipelines that call `gsd auto` and assert no prior session is active — they fail on this exact error and mask the actual cause.

## Impact

- Affected segment: every GSD user running auto-mode on multi-task slices, especially unattended. The bug fires silently; only visible symptom is "no progress" + the cryptic `Session already active` rejection.
- Severity: **medium-high** — the project is unrecoverable through the documented surface; users hit "gsd dead, why?" without realizing state has diverged.
- Frequency: every `orchestrator-terminal / reason: pause` event. In the M032 run alone we have at least 2 such events (`post-unit-finalize-end` seq 5 in iteration 11 triggered advance, and seq 29 triggered the terminal).
- Consequence if not fixed: users will keep hand-editing SQLite or restart daemons, neither of which is documented behavior. Recovery PRs will accumulate on the same root cause.

## Evidence or prior art

- **#414** "auto-mode permanent pause: artifact_db_status_divergence from failure summaries blocks dispatch of pending tasks" — closed. Same architectural surface (artifact↔DB divergence) but the trigger is failure-summary divergence, not orchestrator-terminal divergence. My case is the *pause-reason* variant of the same family — the status-reconciliation path needs to be generalized to cover both triggers.
- **#1243 / #1235** "an exception during auto-start bypasses `cleanupAfterLoopExit`, leaking `s.active` and the auto lock" — closed. Different leak vector (in-memory `s.active` flag), but symptom is the same: `/gsd auto` silently no-ops. Suggests the team has already invested in surface-level cleanup hooks; this issue asks for the persistence-side counterpart.
- **#1485** "Auto-mode leaks SDK tasks in status:running on unit abort/re-dispatch" — **open**, closely related. My bug is the same shape but the leaky subject is the `sessions` row (MCP table), not the SDK `Task` record. Likely the same root cause class — `status` column writes are not transactional with the journal append. Filing this as orthogonal in case #1485's fix only covers the SDK layer.
- **#426** "auto-mode: `lastFinalizedUnitKey` reset at orchestrator.js:259 lets completed units re-dispatch (X→Y→X)" — closed. Different mechanism but illustrates that the orchestrator's in-memory state has been the source of multiple re-dispatch bugs; the divergence between in-memory state and DB state is a recurring theme.

## Additional information

- **Suggested labels:** `bug`, `needs-triage`, `needs-maintainer-review`, `auto-mode`, `state-management`
- **Willing to PR:** yes, with a maintainer's review on the schema. The fix lives in the orchestrator's terminal-event handler and `startSession`'s precheck — small, surgical, with good test coverage potential.
- **Edge cases to consider:** what if the user has another active session in another project pointing at the same DB? The reconciliation should scope by `project_id`/`session_id` pair, not blanket-update.
- **Testing hint:** A test that issues an `orchestrator-terminal / reason: pause` event in isolation, then calls `gsd auto`, then asserts the rejection error message now includes the snapshot's `Blockers`/`Next` lines, would exercise the (1) and (4) fixes end-to-end.
- **Observed during:** M032 metadata_page milestone on `~/java/src/zerone-all`. Specifically, after S02-T01 (新建 `src/api/metadata/page.test.ts` + 修改 `constants.ts` 追加 `PAGE_EDIT` 常量 + 新建 `PageEditorStub.vue` 占位路由) finalized.

### PR DESCRIPTION
**Pull request by jeremymcs**
## Summary
- Added paused session handling for MCP/daemon managers and verified syntax/diff checks; package tests were blocked by missing dependencies and registry ENOTFOUND.

## Verification
- Completed in the repository worktree before push.

## Related Issue
- Closes #1490
- [#1490 [Bug]: gsd-auto-orchestrator terminal/pause event leaves sessions table status='running'; /gsd auto deadlocks indefinitely](https://github.com/open-gsd/gsd-pi/issues/1490)

## User
- @hermer

## Repo
- `open-gsd/gsd-pi`

## Branch
- `issue/1490-bug-gsd-auto-orchestrator-terminal-pause-1784416051`

---

> [!NOTE]
> **Medium Risk**
> Touches core session lifecycle, duplicate-start rules, and Discord bridge teardown paths; behavior changes are well-tested but affect orchestration and long-running daemon sessions.
> 
> **Overview**
> Fixes **#1490** where orchestrator pause/terminal events left sessions stuck as `running`, blocking another `/gsd auto` for the same project.
> 
> **Session lifecycle:** Daemon and MCP `SessionManager` now recognize **paused** (notify text, structured `orchestrator-terminal` / `orchestrator-guard-block` events). Status becomes `paused`, blockers are cleared, and the RpcClient subscription stays so resumed output and later completion still work. **Paused** is treated like other terminal states for duplicate starts—eviction stops the old client and, for paused runs, emits `session:cancelled` so Discord bridge state is not stranded. Shutdown **cleanup** also stops paused child processes.
> 
> **Discord bridge:** `EventBridge` listens for `session:paused`, flushes the message batcher (new public `flush()` on `MessageBatcher`), and stops blocker collectors **without** tearing down channel mappings, so “type to interact” and resumed streaming keep working.
> 
> **Headless helpers:** `headless-events` classifies the same orchestrator pause signals as terminal/blocked where appropriate; tests cover daemon, MCP, bridge, and headless behavior.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit d1ac64dda3a9efcfbfe9a77977dc6191ae8d67e1. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

---
<a href="https://app.blacksmith.sh/open-gsd/codesmith/gsd-pi/pr/1495"><picture><source media="(prefers-color-scheme: dark)" srcset="https://pr-comments-assets.blacksmith.sh/codesmith/view-with-codesmith-dark-v2.svg"><source media="(prefers-color-scheme: light)" srcset="https://pr-comments-assets.blacksmith.sh/codesmith/view-with-codesmith-light-v2.svg">View with Codesmith</picture></a>
<sup>Need help on this PR? Tag <code>/codesmith</code> with what you need. Autofix is enabled.</sup>

### CODE REVIEW

**Comment by jeremymcs** [comment]
🔎 **Verification — 0 of 1 addressed**

Re-checked PR #1495 against [#1490 [Bug]: gsd-auto-orchestrator terminal/pause event leaves sessions table status='running'; /gsd auto deadlocks indefinitely](https://github.com/open-gsd/gsd-pi/issues/1490).

## Subtasks
- [ ] **[Bug]: gsd-auto-orchestrator terminal/pause event leaves sessions table status='running'; /gsd auto deadlocks indefinite** _(pending: Diff only marks in-memory sessions paused from UI notifications; it does not update/reconcile the SQLite sessions row from orchestrator-terminal/pause events.)_

_Re-run from Patchdeck to refresh._

**Comment by jeremymcs** [comment]
**Accepted** — queued for a code change.
**In progress** — applying the accepted fix.
**Verifying** — checking the applied changes.
**Resolved** — addressed in commit `0059e3e`.

**Comment by jeremymcs** [comment]
Started an automated PR update.

Accepted review feedback is being applied.
Progress and final results will be posted in the related review thread(s).

**Comment by jeremymcs** [comment]
Addressed in commit `0059e3e`.

> Responding to [comment by @jeremymcs](https://github.com/open-gsd/gsd-pi/pull/1495#issuecomment-5013314325):
> 🔎 **Verification — 0 of 1 addressed**

Added structured orchestrator pause handling for `orchestrator-terminal` stop/pause and `advance-paused` events, and emitted daemon `session:paused` so EventBridge/headless consumers can reconcile paused sessions instead of leaving them running. I confirmed there is no repo-owned SQLite `sessions` table writer in this worktree; the fix targets the local lifecycle/status path available here.

File: packages/daemon/src/session-manager.ts [hunk 1]
```diff
@@ -48,6 +53,25 @@ function isTerminalNotification(event: Record<string, unknown>): boolean {
...
+    eventType === 'orchestrator-terminal' && name === 'stop' && reason === 'pause'
+  );
+}
+
+function isPausedEvent(event: Record<string, unknown>): boolean {
+  if (isOrchestratorPausedEvent(event)) return true;
+  if (event.type !== 'extension_ui_request' || event.method !== 'notify') return false;
+  const message = String(event.message ?? '').toLowerCase();
+  return PAUSED_PREFIXES.some((prefix) => message.startsWith(prefix));
+}
```
[the remark this answers is not in the document]
Comment (jeremymcs) [reply packages/daemon/src/session-manager.ts:79]: **Accepted** — queued for a code change.
**In progress** — applying the accepted fix.
**Verifying** — checking the applied changes.
**Resolved** — addressed in commit `b4a67d1`.

File: packages/daemon/src/session-manager.ts [hunk 2]
```diff
@@ -48,6 +53,25 @@ function isTerminalNotification(event: Record<string, unknown>): boolean {
...
+function isOrchestratorPausedEvent(event: Record<string, unknown>): boolean {
+  const data = event.data as Record<string, unknown> | undefined;
+  const eventType = String(event.eventType ?? '');
+  const name = String(data?.name ?? '');
+  const reason = String(data?.reason ?? '').toLowerCase();
+  return (
+    eventType === 'orchestrator-guard-block' && name === 'advance-paused'
+  ) || (
+    eventType === 'orchestrator-terminal' && name === 'stop' && reason === 'pause'
+  );
```
[the remark this answers is not in the document]
Comment (jeremymcs) [reply packages/daemon/src/session-manager.ts:70]: **Accepted** — queued for a code change.
**In progress** — applying the accepted fix.
**Verifying** — checking the applied changes.
**Resolved** — addressed in commit `b4a67d1`.

**Comment by jeremymcs** [comment]
Started an automated PR update.

Accepted review feedback is being applied.
Progress and final results will be posted in the related review thread(s).

File: packages/daemon/src/session-manager.ts [hunk 3]
```diff
@@ -412,6 +445,42 @@ export class SessionManager extends EventEmitter {
...
+        blockerMethod: session.pendingBlocker.method,
+      });
+      this.emit('session:blocked', {
+        sessionId: session.sessionId,
+        projectDir: session.projectDir,
+        projectName: session.projectName,
+        blocker: session.pendingBlocker,
+      });
+      return;
+    }
```
[the remark this answers is not in the document]
Comment (jeremymcs) [reply packages/daemon/src/session-manager.ts]: **Accepted** — queued for a code change.
**In progress** — applying the accepted fix.
**Verifying** — checking the applied changes.
**Resolved** — addressed in commit `b4a67d1`.

File: packages/daemon/src/session-manager.ts [hunk 4]
```diff
@@ -412,6 +442,23 @@ export class SessionManager extends EventEmitter {
...
+      session.pendingBlocker = null;
+      session.unsubscribe?.();
+      session.unsubscribe = undefined;
+      this.logger.info('session paused', { sessionId: session.sessionId, projectDir: session.projectDir });
+      this.emit('session:paused', {
+        sessionId: session.sessionId,
+        projectDir: session.projectDir,
+        projectName: session.projectName,
+      });
+      this.markTerminal(session.projectDir);
```
[the remark this answers is not in the document]
Comment (jeremymcs) [reply packages/daemon/src/session-manager.ts]: **Accepted** — queued for a code change.
**In progress** — applying the accepted fix.
**Verifying** — checking the applied changes.
**Resolved** — addressed in commit `b4a67d1`.

File: packages/daemon/src/event-bridge.ts [hunk 5]
```diff
@@ -283,6 +289,15 @@ export class EventBridge {
...
   }
 
+  private async onSessionPaused(data: SessionPausedPayload): Promise<void> {
+    const { sessionId, projectName } = data;
+    // The pause notification has already gone through session:event. Flush any
+    // pending batch and tear down per-session bridge state so a resumed/fresh
+    // run for the same project can register cleanly.
+    await this.cleanupSession(sessionId);
+    this.logger.info('bridge: session paused', { sessionId, projectName });
+  }
```
[the remark this answers is not in the document]
Comment (jeremymcs) [reply packages/daemon/src/event-bridge.ts:308]: **Accepted** — queued for a code change.
**In progress** — applying the accepted fix.
**Verifying** — checking the applied changes.
**Resolved** — addressed in commit `b4a67d1`.

File: packages/daemon/src/session-manager.ts [hunk 6]
```diff
@@ -412,6 +442,22 @@ export class SessionManager extends EventEmitter {
       }
     }
 
+    // Paused detection — pauseAuto() is resumable and must not leave the
```
[the remark this answers is not in the document]
Comment (jeremymcs) [reply packages/daemon/src/session-manager.ts:467]: **Accepted** — queued for a code change.
**In progress** — applying the accepted fix.
**Verifying** — checking the applied changes.
**Resolved** — addressed in commit `b4a67d1`.

**Comment by jeremymcs** [comment]
Started an automated PR update.

Accepted review feedback is being applied.
Progress and final results will be posted in the related review thread(s).

File: packages/daemon/src/session-manager.ts [hunk 7]
```diff
@@ -412,6 +450,22 @@ export class SessionManager extends EventEmitter {
...
+    // session looking active, or duplicate-start prevention will deadlock.
+    if (isPausedEvent(event as Record<string, unknown>)) {
+      session.status = 'paused';
+      session.pendingBlocker = null;
+      session.unsubscribe?.();
+      session.unsubscribe = undefined;
```
[the remark this answers is not in the document]
Comment (jeremymcs) [reply packages/daemon/src/session-manager.ts]: **Accepted** — queued for a code change.
**In progress** — applying the accepted fix.
**Verifying** — checking the applied changes.
**Resolved** — addressed in commit `b4a67d1`.

File: packages/mcp-server/src/session-manager.ts [hunk 8]
```diff
@@ -100,7 +130,7 @@ export class SessionManager {
     const existing = this.sessions.get(resolvedDir);
     if (existing) {
       // Only block when a genuinely active session is running. Terminal
-      // states (error, completed, cancelled) are evicted so the caller can
+      // states (paused, error, completed, cancelled) are evicted so the caller can
```
[the remark this answers is not in the document]
Comment (jeremymcs) [reply packages/mcp-server/src/session-manager.ts:133]: **Accepted** — queued for a code change.
**In progress** — applying the accepted fix.
**Verifying** — checking the applied changes.
**Resolved** — addressed in commit `b4a67d1`.

**Comment by jeremymcs** [comment]
**Accepted** — queued for a code change.
**In progress** — applying the accepted fix.
**Verifying** — checking the applied changes.
**Resolved** — addressed in commit `b4a67d1`.

**Comment by jeremymcs** [comment]
Started an automated PR update.

Accepted review feedback is being applied.
Progress and final results will be posted in the related review thread(s).

File: packages/daemon/src/session-manager.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (jeremymcs) [reply packages/daemon/src/session-manager.ts:79]: Addressed in commit `b4a67d1`.

File: packages/daemon/src/session-manager.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (jeremymcs) [reply packages/daemon/src/session-manager.ts:70]: Addressed in commit `b4a67d1`.

File: packages/daemon/src/session-manager.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (jeremymcs) [reply packages/daemon/src/session-manager.ts]: Addressed in commit `b4a67d1`.

File: packages/daemon/src/session-manager.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (jeremymcs) [reply packages/daemon/src/session-manager.ts]: Addressed in commit `b4a67d1`.

File: packages/daemon/src/event-bridge.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (jeremymcs) [reply packages/daemon/src/event-bridge.ts:308]: Addressed in commit `b4a67d1`.

File: packages/daemon/src/session-manager.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (jeremymcs) [reply packages/daemon/src/session-manager.ts:467]: Addressed in commit `b4a67d1`.

File: packages/daemon/src/session-manager.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (jeremymcs) [reply packages/daemon/src/session-manager.ts]: Addressed in commit `b4a67d1`.

Confirmed both session managers keep the RpcClient event subscription on pause, and added MCP regression coverage proving a paused session still processes later terminal output.

File: packages/mcp-server/src/session-manager.ts (hunk 8 above)
[the remark this answers is not in the document]
Comment (jeremymcs) [reply packages/mcp-server/src/session-manager.ts:133]: Addressed in commit `b4a67d1`.

Confirmed MCP startSession stops evicted paused/terminal clients; aligned the MCP test double and asserted evicted session clients are stopped before a fresh session starts.

**Comment by jeremymcs** [comment]
Addressed in commit `b4a67d1`.

> Responding to [comment by @cursor[bot]](https://github.com/open-gsd/gsd-pi/pull/1495#pullrequestreview-4729615390):
> 

Validated both Bugbot findings and added focused MCP regression coverage for pause subscription retention and paused-session process reclamation.

File: packages/daemon/src/session-manager.ts [hunk 9]
```diff
@@ -310,7 +340,15 @@ export class SessionManager extends EventEmitter {
...
-      if (session.status === 'running' || session.status === 'starting' || session.status === 'blocked') {
+      // A paused session still owns a live headless child process (its RpcClient
+      // is retained for "type to interact"). On daemon shutdown that process must
+      // be reclaimed too, otherwise pausing then stopping the daemon leaks it.
+      if (
+        session.status === 'running' ||
+        session.status === 'starting' ||
+        session.status === 'blocked' ||
+        session.status === 'paused'
+      ) {
```
[the remark this answers is not in the document]
Comment (jeremymcs) [reply packages/daemon/src/session-manager.ts:351]: **Accepted** — queued for a code change.
**In progress** — applying the accepted fix.
**Verifying** — checking the applied changes.
**Resolved** — addressed in commit `d1ac64d`.

**Comment by jeremymcs** [comment]
Started an automated PR update.

Accepted review feedback is being applied.
Progress and final results will be posted in the related review thread(s).

File: packages/daemon/src/session-manager.ts [hunk 10]
```diff
@@ -357,9 +395,31 @@ export class SessionManager extends EventEmitter {
...
       // kept for follow-up relay) — reclaim it on eviction.
       void session.client.stop().catch(() => { /* swallow */ });
+
+      // Paused sessions keep EventBridge channel mappings alive; notify listeners
+      // so batchers, collectors, and channel maps are torn down before restart.
+      this.emit('session:cancelled', {
+        sessionId: session.sessionId,
+        projectDir: session.projectDir,
+        projectName: session.projectName,
+      });
```
[the remark this answers is not in the document]
Comment (jeremymcs) [reply packages/daemon/src/session-manager.ts]: **Accepted** — queued for a code change.
**In progress** — applying the accepted fix.
**Verifying** — checking the applied changes.
**Resolved** — addressed in commit `d1ac64d`.

**Comment by jeremymcs** [comment]
**Accepted** — queued for a code change.
**In progress** — applying the accepted fix.
**Verifying** — checking the applied changes.
**Resolved** — addressed in commit `d1ac64d`.

**Comment by jeremymcs** [comment]
Started an automated PR update.

Accepted review feedback is being applied.
Progress and final results will be posted in the related review thread(s).

File: packages/daemon/src/session-manager.ts (hunk 9 above)
[the remark this answers is not in the document]
Comment (jeremymcs) [reply packages/daemon/src/session-manager.ts:351]: Addressed in commit `d1ac64d`.

File: packages/daemon/src/session-manager.ts (hunk 10 above)
[the remark this answers is not in the document]
Comment (jeremymcs) [reply packages/daemon/src/session-manager.ts]: Addressed in commit `d1ac64d`.

Confirmed `evictSession` only emits `session:cancelled` for paused sessions, and added regression coverage that completed/error sessions do not emit on eviction and cancelled sessions are not emitted a second time.

**Comment by jeremymcs** [comment]
Addressed in commit `d1ac64d`.

> Responding to [comment by @cursor[bot]](https://github.com/open-gsd/gsd-pi/pull/1495#pullrequestreview-4729653903):
> 

Addressed the Bugbot review item by tightening daemon session-manager tests around non-paused terminal eviction behavior; the guarded production implementation was already present in the worktree.
